openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,435 @@
+type managementCPUsOverride struct {
+	*admission.Handler
+	nsLister       corev1listers.NamespaceLister
+	nsListerSynced func() bool
+	nodeLister     corev1listers.NodeLister
+	nodeListSynced func() bool
+}
+
+func (a *managementCPUsOverride) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	a.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+	a.nsListerSynced = kubeInformers.Core().V1().Namespaces().Informer().HasSynced
+	a.nodeLister = kubeInformers.Core().V1().Nodes().Lister()
+	a.nodeListSynced = kubeInformers.Core().V1().Nodes().Informer().HasSynced
+}
+
+func (a *managementCPUsOverride) ValidateInitialization() error {
+	if a.nsLister == nil {
+		return fmt.Errorf("%s did not get a namespace lister", PluginName)
+	}
+	if a.nsListerSynced == nil {
+		return fmt.Errorf("%s plugin needs a namespace lister synced", PluginName)
+	}
+	if a.nodeLister == nil {
+		return fmt.Errorf("%s did not get a node lister", PluginName)
+	}
+	if a.nodeListSynced == nil {
+		return fmt.Errorf("%s plugin needs a node lister synced", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	workloadAnnotations := getWorkloadAnnotations(pod.Annotations)
+
+	// strip any resource annotations specified by a user
+	stripResourcesAnnotations(pod.Annotations)
+	// strip any workload annotation to prevent from underlying components(CRI-O, kubelet) to apply any changes
+	// according to the workload annotation
+	stripWorkloadAnnotations(pod.Annotations)
+
+	// more than one workload annotation exists under the pod and we do not support different workload types
+	// under the same pod
+	if len(workloadAnnotations) > 1 {
+		pod.Annotations[workloadAdmissionError] = fmt.Sprintf(admissionErrorMultipleWorkloadAnnotations, workloadAnnotations)
+		return nil
+	}
+
+	workloadType, err := getWorkloadType(workloadAnnotations)
+	if err != nil {
+		pod.Annotations[workloadAdmissionError] = fmt.Sprintf(admissionErrorFailedGetWorkloadType, err)
+		return nil
+	}
+
+	// no workload annotation is available
+	if workloadType == "" {
+		return nil
+	}
+
+	workloadAnnotation := fmt.Sprintf("%s%s", podWorkloadAnnotationPrefix, workloadType)
+	// TODO: currently we do not check the effect, but it can be useful in the future
+	_, err = getWorkloadAnnotationEffect(workloadAnnotations[workloadAnnotation])
+	if err != nil {
+		pod.Annotations[workloadAdmissionError] = fmt.Sprintf(admissionErrorFailedGetWorkloadAnnotationEffect, err)
+		return nil
+	}
+
+	if !a.waitForSyncedStore(time.After(timeToWaitForCacheSync)) {
```

Comment:
this must be capped by the context timeout. Otherwise you will see 500 errors in the API and not this pretty error.

Reply:
do you mean something like
```
for !a.nsListerSynced() || !a.nodeListSynced() {
		select {
		case <-time.After(100 * time.Millisecond):
		case <-ctx.Done:
			return a.nsListerSynced() && a.nodeListSynced()
		}
	}
```

Reply:
`timeToWait := min(contextTimeout, timeToWaitForCacheSync)`

Reply:
Other admission plugin don't do this either. Not super critical.